### DISCUSSION TITLE
Automatic suggestions for `# type: ignore[xxx]`

### DISCUSSION THREAD

**Discussion by vors** [microsoft/pyright#4712] [General]
I'm migrating from mypy to pyright. One of the ergonomics that I'm really missing is ability to produce the PEP484 style supressions: `# type: ignore[attr]` for example. I mean the `attr` part -- it would be good to explicitly tell how to scope the supression only to the applicable mnemonics in a type-checker-agnostic way.

Thoughts on it? Maybe I'm missing some settings or some point?

**Comment by erictraut** [comment]
There is no way to suppress specific diagnostics in a type-checker-agnostic way because diagnostic rules are not standardized across type checkers. 

You can use `# type: ignore` to suppress all diagnostics on a line (as specified in PEP 484). Pyright also supports `# pyright: ignore[xxx]` for pyright-specific diagnostic suppression.

See [this documentation](https://github.com/microsoft/pyright/blob/main/docs/comments.md#line-level-diagnostic-suppression) for more details.

Mypy's use of brackets after `# type: ignore` is non-standard (not specified in PEP 484), and there has been some discussion of adding a `# mypy: ignore[xxx]` to suppress mypy-specific diagnostics. You may be interested in [this discussion](https://github.com/python/mypy/issues/12358).

**Comment by vors** [comment]
Oh I see, thank you for the explanation!

Just want to double-check, when you say

> Pyright also supports # pyright: ignore[xxx] for pyright-specific diagnostic suppression.

Do you mean that `# type: ignore[reportOptionalMemberAccess]` would suppress only `reportOptionalMemberAccess` or is that `# type: ignore[reportOptionalMemberAccess]` is a supported syntax but would still suppress everything?

It seems to be the former, because of this example:

```
a = None
if 1 == 2:
    a = 123
1 + "123" + a.x  # type: ignore[reportOptionalMemberAccess]
```

This doesn't produce ANY errors, but if I remove the `# type: ignore[reportOptionalMemberAccess]` part then it produces 2 different types of errors
```
dir/foo.py
  dir/foo.py:4:1 - error: Operator "+" not supported for types "Literal[1]" and "Literal['123']"
    Operator "+" not supported for types "Literal[1]" and "Literal['123']" (reportGeneralTypeIssues)
  dir/foo.py:4:15 - error: "x" is not a known member of "None" (reportOptionalMemberAccess)
  dir/foo.py:4:15 - error: Cannot access member "x" for type "Literal[123]"
    Member "x" is unknown (reportGeneralTypeIssues)
  dir/foo.py:4:1 - warning: Expression value is unused (reportUnusedExpression)
```

**Reply by erictraut** [reply]
`# type: ignore` suppresses all errors on the line regardless of whether the comment is followed by brackets. This is the behavior specified by PEP 484, which says nothing about optional brackets after `# type: ignore`. Mypy's behavior here does not match PEP 484. When mypy's maintainers added this support, they were probably not thinking about other type checkers.

If you want to suppress specific pyright errors, you can use `# pyright: ignore[xxx]`. This is a pyright-specific mechanism, so it doesn't violate the PEP 484 standard.

Ideally, mypy would add a mypy-specific ignore format, like `# mypy: ignore[attr]`. That way, you could combine ignore comments in cases where you want to use multiple type checkers on a code base (e.g. `# mypy: ignore[attr] # pyright: ignore[reportGeneralTypeIssues]`).

**Comment by vors** [comment]
Got it, thank you! I will probably lean toward using `# pyright: ignore[xxx]` for new suppressions then.
